File: test/integ/BitwiseOpsConstantPropagationTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@Suppress("HexColorValueUsage", "ClassNameDoesNotMatchFileName")
class TestRGBA {
  /** Extracts the green component from a 32-bit RGBA value. */
  inline fun extractGreenFromRGBA256(rgba: Int): Int {
    return (rgba ushr 16) and 0xFF
  }

  // Test >>>, |, & on int.
  public fun mainExtractGreen(baseColor: Int) {
    var rgba256 = baseColor or 0x00F10000
    rgba256 = extractGreenFromRGBA256(rgba256)
    if ((rgba256 and 0xF0) != 0) {
      print("8-bit deep green")
    } else {
      // This branch should be optimized out.
      print("8-bit light green")
    }
  }

  /** Extracts the green component from a 64-bit RGBA value. */
  inline fun extractGreenFromRGBA512(rgba: Long): Long {
    return (rgba ushr 32) and 0xFFFF
  }

  // Test >>>, |, & on long.
  public fun mainExtractGreen(baseColor: Long) {
    var rgba512: Long = baseColor or 0x0000010000000000L
    rgba512 = extractGreenFromRGBA512(rgba512)
    if ((rgba512 and 0xFF00L) != 0L) {
      print("16-bit deep green")
    } else {
      // This branch should be optimized out.
      print("16-bit light green")
    }
  }

  /** Tells if rgba has any non-red component has all. */
  inline fun hasNonRed(rgba: Int): Boolean {
    return (rgba shl 8) != 0
  }

  // Test <<, & on int.
  public fun mainHasNonRed(baseColor: Int) {
    val onlyLowerRed = baseColor and 0x0F000000
    if (hasNonRed(onlyLowerRed)) {
      // This branch should be optimized out.
      print("int onlyLowerRed has non-red")
    } else {
      print("int onlyLowerRed has no non-red")
    }

    val onlyLowerAlpha = baseColor and 0xF
    // Neither branch should be optimized out.
    if (hasNonRed(onlyLowerAlpha)) {
      print("int onlyLowerAlpha has non-red")
    } else {
      print("int onlyLowerAlpha has no non-red")
    }
  }

  /** Tells if rgba has any non-red component has all. */
  inline fun hasNonRed(rgba: Long): Boolean {
    return (rgba shl 16) != 0L
  }

  // Test <<, & on long.
  public fun mainHasNonRed(baseColor: Long) {
    val onlyLowerRed = baseColor and 0x00FF000000000000L
    if (hasNonRed(onlyLowerRed)) {
      // This branch should be optimized out.
      print("long onlyLowerRed has non-red")
    } else {
      print("long onlyLowerRed has no non-red")
    }

    val onlyLowerAlpha = baseColor and 0xFFL
    // Neither branch should be optimized out.
    if (hasNonRed(onlyLowerAlpha)) {
      print("long onlyLowerAlpha has non-red")
    } else {
      print("long onlyLowerAlpha has no non-red")
    }
  }

  inline fun invert(rgba: Int): Int {
    return rgba.inv()
  }

  inline fun invertAlpha(rgba: Int): Int {
    return rgba xor 0xFF
  }

  // Test ~, ^, & on int
  public fun mainInvert(baseColor: Int) {
    val alphaless = baseColor and 0xFF.inv()
    var inverted = invert(alphaless) // inverted lowest 8 bits are 1s
    if (inverted == 0) {
      // This branch should be optimized out
      print("int alphaless inverted is zero")
    } else {
      print("int alphaless inverted is not zero")
    }

    // Neither branch should be optimized out
    if (inverted == 0xFF) {
      print("int alphaless inverted is 0xFF")
    } else {
      print("int alphaless inverted is not 0xFF")
    }

    inverted = invertAlpha(inverted) // inverted lowest 8 bits are 0s
    // Neither branch should be optimized out
    if (inverted == 0) {
      print("int alphaless inverted twice is zero")
    } else {
      print("int alphaless inverted twice is not zero")
    }

    if (inverted == 0xFF) {
      // This branch should be optimized out
      print("int alphaless inverted twice is 0xFF")
    } else {
      print("int alphaless inverted twice is not 0xFF")
    }
  }

  inline fun invert(rgba: Long): Long {
    return rgba.inv()
  }

  inline fun invertAlpha(rgba: Long): Long {
    return rgba xor 0xFFFFL
  }

  // Test ~, & on long
  public fun mainInvert(baseColor: Long) {
    val alphaless = baseColor and 0xFFFFL.inv()
    var inverted = invert(alphaless) // inverted lowest 16 bits are 1s
    if (inverted == 0L) {
      // This branch should be optimized out
      print("long alphaless inverted is zero")
    } else {
      print("long alphaless inverted is not zero")
    }

    // Neither branch should be optimized out
    if (inverted == 0xFFFFL) {
      print("long alphaless inverted is 0xFFFF")
    } else {
      print("long alphaless inverted is not 0xFFFF")
    }

    inverted = invertAlpha(inverted) // inverted lowest 16 bits are 0s
    // Neither branch should be optimized out
    if (inverted == 0L) {
      print("long alphaless inverted twice is zero")
    } else {
      print("long alphaless inverted twice is not zero")
    }

    if (inverted == 0xFFL) {
      // This branch should be optimized out
      print("long alphaless inverted twice is 0xFFFF")
    } else {
      print("long alphaless inverted twice is not 0xFFFF")
    }
  }
}

File: test/integ/BitwiseOpsConstantPropagationTest.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <string_view>

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include "ConstantPropagationPass.h"
#include "DexUtil.h"
#include "RedexTest.h"

using ::testing::AllOf;
using ::testing::HasSubstr;
using ::testing::Not;

class BitwiseOpsConstantPropagationTest : public RedexIntegrationTest {
 protected:
  void set_root_method(std::string_view full_name) {
    auto* method = DexMethod::get_method(full_name)->as_def();
    ASSERT_NE(nullptr, method) << "Could not find method " << full_name;
    method->rstate.set_root();
  }
};
namespace {

TEST_F(BitwiseOpsConstantPropagationTest, ExtractGreenInt) {
  static constexpr std::string_view root_method_name =
      "LTestRGBA;.mainExtractGreen:(I)V";

  auto scope = build_class_scope(stores);

  set_root_method(root_method_name);

  Pass* constp = new ConstantPropagationPass();
  std::vector<Pass*> passes = {constp};
  run_passes(passes);

  auto* const method = DexMethod::get_method(root_method_name)->as_def();
  auto* const code = method->get_code();
  ASSERT_NE(nullptr, code);
  const auto code_str = assembler::to_string(code);
  EXPECT_THAT(code_str, HasSubstr("8-bit deep green"))
      << "\"8-bit deep green\" branch is optimized out, but should't";
  EXPECT_THAT(code_str, Not(HasSubstr("8-bit light green")))
      << "\"8-bit light green\" should be optimized out, but isn't";
}

TEST_F(BitwiseOpsConstantPropagationTest, ExtractGreenLong) {
  static constexpr std::string_view root_method_name =
      "LTestRGBA;.mainExtractGreen:(J)V";

  auto scope = build_class_scope(stores);

  set_root_method(root_method_name);

  Pass* constp = new ConstantPropagationPass();
  std::vector<Pass*> passes = {constp};
  run_passes(passes);

  auto* const method = DexMethod::get_method(root_method_name)->as_def();
  auto* const code = method->get_code();
  ASSERT_NE(nullptr, code);
  const auto code_str = assembler::to_string(code);
  EXPECT_THAT(code_str, HasSubstr("16-bit deep green"))
      << "\"16-bit deep green\" branch is optimized out, but should't";
  EXPECT_THAT(code_str, Not(HasSubstr("16-bit light green")))
      << "\"16-bit light green\" is unexpectedly not optimized out";
}

TEST_F(BitwiseOpsConstantPropagationTest, HasNonRedInt) {
  static constexpr std::string_view root_method_name =
      "LTestRGBA;.mainHasNonRed:(I)V";

  auto scope = build_class_scope(stores);

  set_root_method(root_method_name);

  Pass* constp = new ConstantPropagationPass();
  std::vector<Pass*> passes = {constp};
  run_passes(passes);

  auto* const method = DexMethod::get_method(root_method_name)->as_def();
  auto* const code = method->get_code();
  ASSERT_NE(nullptr, code);
  const auto code_str = assembler::to_string(code);
  EXPECT_THAT(code_str, Not(HasSubstr("int onlyLowerRed has non-red")))
      << "\"int onlyLowerRed has non-red\" is unexpectedly not optimized out";
  EXPECT_THAT(code_str, HasSubstr("int onlyLowerRed has no non-red"))
      << "\"int onlyLowerRed has no non-red\" branch is unexpectedly optimized "
         "out";
  EXPECT_THAT(code_str,
              AllOf(HasSubstr("int onlyLowerAlpha has non-red"),
                    HasSubstr("int onlyLowerAlpha has no non-red")))
      << "either int onlyLowerAlpha branch is unexpectedly optimized out";
}

TEST_F(BitwiseOpsConstantPropagationTest, HasNonRedLong) {
  static constexpr std::string_view root_method_name =
      "LTestRGBA;.mainHasNonRed:(J)V";

  auto scope = build_class_scope(stores);

  set_root_method(root_method_name);

  Pass* constp = new ConstantPropagationPass();
  std::vector<Pass*> passes = {constp};
  run_passes(passes);

  auto* const method = DexMethod::get_method(root_method_name)->as_def();
  auto* const code = method->get_code();
  ASSERT_NE(nullptr, code);
  const auto code_str = assembler::to_string(code);
  EXPECT_THAT(code_str, Not(HasSubstr("long ohnlyLowerRed has non-red")))
      << "\"long onlyLowerRed has non-red\" should be optimized out, but isn't";
  EXPECT_THAT(code_str, HasSubstr("long onlyLowerRed has no non-red"))
      << "\"long onlyLowerRed has no non-red\" branch is unexpectedly "
         "optimized out";
  EXPECT_THAT(code_str,
              AllOf(HasSubstr("long onlyLowerAlpha has non-red"),
                    HasSubstr("long onlyLowerAlpha has no non-red")))
      << "either long onlyLowerAlpha branch is unexpectedly optimized out";
}

TEST_F(BitwiseOpsConstantPropagationTest, InvertInt) {
  static constexpr std::string_view root_method_name =
      "LTestRGBA;.mainInvert:(I)V";

  auto scope = build_class_scope(stores);

  set_root_method(root_method_name);

  Pass* constp = new ConstantPropagationPass();
  std::vector<Pass*> passes = {constp};
  run_passes(passes);

  auto* const method = DexMethod::get_method(root_method_name)->as_def();
  auto* const code = method->get_code();
  ASSERT_NE(nullptr, code);
  const auto code_str = assembler::to_string(code);
  EXPECT_THAT(code_str, Not(HasSubstr("int alphaless inverted is zero")))
      << "\"int alphaless inverted is zero\" is unexpectedly not optimized out";
  EXPECT_THAT(code_str, HasSubstr("int alphaless inverted is not zero"))
      << "\"int alphaless inverted is not zero\" branch is unexpectedly "
         "optimized out";
  EXPECT_THAT(code_str,
              AllOf(HasSubstr("int alphaless inverted is 0xFF"),
                    HasSubstr("int alphaless inverted is not 0xFF")))
      << "either \"int alphaless inverted is/isn't 0xFF branch\" is "
         "unexpectedly optimized out";
  EXPECT_THAT(code_str,
              AllOf(HasSubstr("int alphaless inverted twice is zero"),
                    HasSubstr("int alphaless inverted twice is not zero")))
      << "either \"int alphaless inverted twice is/isn't zero\" branch is "
         "unexpectedly optimized out";
  EXPECT_THAT(code_str, Not(HasSubstr("int alphaless inverted twice is 0xFF")))
      << "\"int alphaless inverted twice is 0xFF\" is unexpectedly not "
         "optimized out";
  EXPECT_THAT(code_str, HasSubstr("int alphaless inverted twice is not 0xFF"))
      << "\"int alphaless inverted twice is not 0xFF\" branch is unexpectedly "
         "optimized out";
}

TEST_F(BitwiseOpsConstantPropagationTest, InvertLong) {
  static constexpr std::string_view root_method_name =
      "LTestRGBA;.mainInvert:(J)V";

  auto scope = build_class_scope(stores);

  set_root_method(root_method_name);

  Pass* constp = new ConstantPropagationPass();
  std::vector<Pass*> passes = {constp};
  run_passes(passes);

  auto* const method = DexMethod::get_method(root_method_name)->as_def();
  auto* const code = method->get_code();
  ASSERT_NE(nullptr, code);
  const auto code_str = assembler::to_string(code);
  EXPECT_THAT(code_str, Not(HasSubstr("long alphaless inverted is zero")))
      << "\"long alphaless inverted is zero\" is unxpectedly not optimized out";
  EXPECT_THAT(code_str, HasSubstr("long alphaless inverted is not zero"))
      << "\"long alphaless inverted is not zero\" branch is unexpectedly "
         "optimized out";
  EXPECT_THAT(code_str,
              AllOf(HasSubstr("long alphaless inverted is 0xFFFF"),
                    HasSubstr("long alphaless inverted is not 0xFFFF")))
      << "either \"long alphaless inverted is/isn't 0xFFFF branch\" is "
         "unexpectedly optimized out";
  EXPECT_THAT(code_str,
              AllOf(HasSubstr("long alphaless inverted twice is zero"),
                    HasSubstr("long alphaless inverted twice is not zero")))
      << "either \"long alphaless inverted twice is/isn't zero\" branch is "
         "unexpectedly optimized out";
  EXPECT_THAT(code_str,
              Not(HasSubstr("long alphaless inverted twice is 0xFFFF")))
      << "\"long alphaless inverted twice is 0xFFFF\" is unexpectedly not "
         "optimized out";
  EXPECT_THAT(code_str,
              HasSubstr("long alphaless inverted twice is not 0xFFFF"))
      << "\"long alphaless inverted twice is not 0xFFFF\" branch is "
         "unexpectedly optimized out";
}
} // namespace
